### PR TITLE
[search] Support specific field search

### PR DESCRIPTION
**Pull request by pmphfm**
Also adds support for searching extended tags
Fixes tests cases so that it is more readable and easily extendable.

Test: Adds unit and integration tests.

### CODE REVIEW

**Comment by Saecki** [comment]
Wouldn't it be easier to send the specific search queries inside the body, or as proper url parameters?

**Comment by pmphfm** [comment]
Yes that also works but that is just lot more effort (as every client needs to implement parsing and building query). This puts it in one place(leading to a sort of uniformity across clients).

That said, I am using this for a while and thought others might find it useful to have it now.  This is how android app looks and web client code is unchanged.

Screenshot

**Comment by Saecki** [comment]
Oh got it, so your query syntax is meant to be used by the user directly.

While that simplifies a bit inside the client at first, I personally feel like this stringly typed syntax would belong in the client. When extending the search functionality of a client to support multiple text fields to search for specific properties, this syntax would be ambiguous for a number of inputs. Say we search for a artist's name containing an apostrophe. Since the apostrophes that contain the value are already percent encoded, the one inside the search input would act as a delimiter or would have to be escaped in another way.

**Comment by pmphfm** [comment]
Yes it is pretty much like the Gmail's search operator. And like that feature, you can still use regular search if you have special characters. Basically for all most all cases this gives pretty powerful interface and for some specific cases you still get to use old interface.

**Comment by agersant** [comment]
Hello!

I haven't had time yet to look at the implementation for this but I did follow the conversation. I think this approach is the right direction. I like having the server being responsible for parsing the search query because:
- it makes it easier to implement search in new clients
- it guarantees search will will behave consistently between clients

**Comment by micwoj92** [comment]
This branch has conflicts that must be resolved

**Comment by pmphfm** [comment]
Hi @micwoj92, 

tldr: I am maintaining a downstream [fork](https://github.com/pmphfm/polaris) for lack of interests/priority for the feature.

I have slightly higher expectations(than what the original author has in mind) from a music streaming service as I have moved to rust based service from other preexisting self hosted services. Some of these expectations tie into features like
- First class search support
- Better playlist load [support](https://github.com/pmphfm/polaris/blob/master/docs/user-guide-src/src/playlists.md#load)
- [Export](https://github.com/pmphfm/polaris/blob/master/docs/user-guide-src/src/playlists.md#export)/import m3u playlist for portability across services. I want to bring in playlist created over years and in future I want to be free to take the data out of polaris
- Smart playlists
- [RJ mode](https://github.com/pmphfm/polaris/commit/ec06245e2f5e74d69632988e425f9add689e2ed4#diff-59e2ee6068bcebac2b80ba4f4bdb951840064e1f416c255a3fec985da1ffd841)

Some of these are already pushed to the github fork, including [front-end](https://github.com/pmphfm/polaris-web/commits/master) changes, but some are still in my local branch. Please check it out if that interests you.

Thanks,

**Comment by agersant** [comment]
Very happy to see an alive and well fork 🥳 

I work on Polaris at a pretty glacial pace (as a hobby, when I have time and energy) so it makes me happy if it can be used to build something that helps more people! I love the playlist improvements and m3u support. Thanks for posting here!

One thing I would request (if you are planning to get more users/contributors on the fork) is to differentiate the name/logo so different flavors of Polaris don't easily get mistaken for one another.

**Comment by agersant** [comment]
Implemented in polaris 0.15.0 🥳
